src/AzureIoTHub.Portal.Infrastructure/Jobs/AWS/SyncGreenGrassDevicesJob.cs

```diff
@@ -0,0 +1,199 @@
+                if (edgeDeviceModel == null)
+                {
+                    this.logger.LogWarning($"Cannot import Greengrass device '{thing.ThingName}'. The EdgeDeviceModel '{thing.ThingTypeName}' doesn't exist");
+                    continue;
+                }
+
+                //Map with EdgeDevice
+                var edgeDevice = this.mapper.Map<EdgeDevice>(thing);
+                edgeDevice.DeviceModelId = edgeDeviceModel.Id;
+                //EdgeDevices properties that are not present in the thing
+                try
+                {
+                    var modules = await this.configService.GetConfigModuleList(edgeDevice.DeviceModelId);
+                    edgeDevice.NbDevices = await this.awsExternalDevicesService.GetEdgeDeviceNbDevices(this.mapper.Map<IoTEdgeDevice>(edgeDevice));
+                    edgeDevice.NbModules = modules.Count;
+                    var coreDevice = await amazonGreenGrass.GetCoreDeviceAsync(new GetCoreDeviceRequest() { CoreDeviceThingName = thing.ThingName });
+                    if (coreDevice.HttpStatusCode != HttpStatusCode.OK)
+                    {
+                        this.logger.LogWarning($"Cannot import Greengrass device '{thing.ThingName}' due to an error retrieving core device in the Amazon IoT Data API : {coreDevice.HttpStatusCode}");
+                        continue;
+                    }
+                    edgeDevice.ConnectionState = coreDevice.Status == CoreDeviceStatus.HEALTHY ? "Connected" : "Disconnected";
+                }
+                catch (Exception e)
+                {
+                    this.logger.LogWarning($"Cannot import Greengrass device '{thing.ThingName}' due to an error retrieving Greengrass device properties in the Amazon IoT Data API.", e);
+                    continue;
+                }
+
+                //Create or update the Edge Device
+                await CreateOrUpdateGreenGrassDevice(edgeDevice);
+            }
+
+            foreach (var item in (await this.edgeDeviceRepository.GetAllAsync(
+                edgeDevice => !things.Select(x => x.ThingId).Contains(edgeDevice.Id),
+                default,
+                d => d.Tags,
+                d => d.Labels
+            )))
+            {
+                this.edgeDeviceRepository.Delete(item.Id);
+            }
+
+            await this.unitOfWork.SaveAsync();
+        }
+
+        private async Task<List<DescribeThingResponse>> GetAllGreenGrassDevices()
+        {
+            var devices = new List<DescribeThingResponse>();
+
+            var nextToken = string.Empty;
+
+            var response = await amazonGreenGrass.ListCoreDevicesAsync(
+                new ListCoreDevicesRequest
+                {
+                    NextToken = nextToken
+                });
+
+            foreach (var requestDescribeThing in response.CoreDevices.Select(device => new DescribeThingRequest { ThingName = device.CoreDeviceThingName }))
```

Comment:
2 things:
- you should loop on every pages here
- I would prefer to sync devices (not core devices)

It might be preferable to look for devices directly instead of getting core devices that might not be present until the edge is well connected and send back to AWS its status...
Furthermore, it might help to reduce requests to AWS by only getting the list of devices with their thingType.